pkg/operator/latencyprofilecontroller/controller.go

```diff
@@ -0,0 +1,216 @@
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	v1 "github.com/openshift/client-go/config/clientset/versioned/typed/config/v1"
+	listerv1 "github.com/openshift/client-go/config/listers/config/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	configv1client "github.com/openshift/client-go/config/clientset/versioned"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	controlplanev1 "github.com/openshift/api/kubecontrolplane/v1"
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/operatorclient"
+)
+
+const (
+	operatorNamespace = operatorclient.OperatorNamespace
+	targetNamespace   = operatorclient.TargetNamespace
+
+	ApiServerConfigMapName = "config"
+	ApiServerConfigMapKey  = "config.yaml"
+
+	DefaultNotReadyTolerationSecondsArgument    = "default-not-ready-toleration-seconds"
+	DefaultUnreachableTolerationSecondsArgument = "default-unreachable-toleration-seconds"
+)
+
+// LatencyProfileController
+type LatencyProfileController struct {
+	operatorClient  v1helpers.StaticPodOperatorClient
+	configMapClient corev1client.ConfigMapsGetter
+	nodeLister      listerv1.NodeLister
+	nodeClient      v1.NodeInterface
+}
+
+func NewLatencyProfileController(
+	operatorClient v1helpers.StaticPodOperatorClient,
+	configClient configv1client.Clientset,
+	configInformer configinformers.SharedInformerFactory,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	kubeClient kubernetes.Interface,
+	eventRecorder events.Recorder,
+) factory.Controller {
+
+	ret := &LatencyProfileController{
+		operatorClient:  operatorClient,
+		configMapClient: v1helpers.CachedConfigMapGetter(kubeClient.CoreV1(), kubeInformersForNamespaces),
+		nodeLister:      configInformer.Config().V1().Nodes().Lister(),
+		nodeClient:      configClient.ConfigV1().Nodes(),
+	}
+	configClient.ConfigV1().Nodes()
+
+	return factory.New().WithInformers(
+		kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().ConfigMaps().Informer(),
+		operatorClient.Informer(),
+		configInformer.Config().V1().Nodes().Informer(),
+	).ResyncEvery(time.Second).WithSync(ret.sync).ToController("LatencyProfileController", eventRecorder)
+}
+
+func (c *LatencyProfileController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	syncMethod := c.ensureProfileSyncedToAllNodes
+	err := syncMethod(ctx, syncCtx)
+	utilruntime.HandleError(err)
+	return err
+}
+
+// ensureProfileSyncedToAllNodes
+func (c *LatencyProfileController) ensureProfileSyncedToAllNodes(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
could you please explain the purpose of this method?

Reply:
Having the observers is not enough?

Reply:
We'd want to add the latency profile controller apart from just the config observers in favour of the following desired outcomes:
- having a controller would ensure us to update the config node CR status condition with appropriate status of K-apiserver
- we have certain edge cases where we would like the user refrain the user from transitioning latency profiles to/from extremeties, having a controller would allow rejecting such use cases back in the config observer
- we want to ensure that all control plane nodes in the cluster have reached the desired state i.e. applied the required K-apiserver flags before another update is trigerred via config observers, this helps in cluster to stay functional/usable

More: https://coreos.slack.com/archives/C0392N52Q5P/p1649867165534869